### ISSUE TITLE
Change cost function API

### ISSUE THREAD

**Issue by tinosulzer** [paramm-team/pybamm-param#37]
Currently, you have to pass in `data`, variables_optimize`, and `cost_function` separately to the `DataFit` problem.

They could all be rolled into a single input `cost_function` which takes in a `pybamm.Solution` and returns an error scalar or array. Something like
```python3
def cost_full(solution, data, variables_optimize):
   total = 0
   for var in variables_optimize:
      total += np.sum((data[var] - solution[var].data) ** 2)
   return total

cost_function = functools.partial(cost_full, data=data, variables_optimize=["Terminal voltage [V]"]) # this is now a function of 'solution' only
```

This is a better separation of concerns and gives the user full flexibility to specify a cost function, weights, etc. Then you can have helper functions to construct the appropriate cost function using RMSE error, MLE, etc.

If you like this suggestion, I'm happy to implement this to get myself more familiar with the package.

**Comment by brosaplanella** [comment]
I am not sure I follow you. Isn't this just moving where different inputs are requested?

The overall idea atm (and I will hopefully have some diagrams for it later in the week) is that an `OptimisationProblem` (for the moment we have `DataFit` and `OCPBalance`) which is analogous to a PyBaMM `Simulation` so takes all the necessary bits for it to work and it is fairly straightforward.

Note that `DataFit` is probably the most generic kind of problem we want to solve. The structure should work across various problems (so, even if they are not implemented, it is useful to think how `GITT` or `EIS` classes would work into this). For example, for GITT, I would expect users to choose an approach and then the package assembles the model for you (e.g. `a *sqrt(t) + b` or SPM directly).

Happy to discuss that further!

**Comment by tinosulzer** [comment]
It is just moving inputs to a different place but it's about giving users access to them.
Basically I am just saying the [full optimization function](https://github.com/paramm-team/pybamm-param/blob/ce8de11a510fc5a2ca02dff9eaa1318f52e3ef30/pbparam/optimisation_problems/data_fit.py#L36) should be an input to `DataFit`, instead of `variables_optimize` and `cost`. That way if a user wants to edit the function in some way (e.g. add weights), they can just do it locally instead of having to edit `data_fit.py`

**Comment by brosaplanella** [comment]
I see what you mean. Is there a way to keep the simple API but allow for customisation? I would really like to keep an API like
```python3
ocp_balance = pbparam.OCPBalance(
    [anode_three_lit, anode_three_delit], [anode_half_lit, anode_half_delit]
)
optimiser = pbparam.ScipyMinimize(method="Nelder-Mead")
result = optimiser.optimise(ocp_balance)
```
I am using `OCPBalance` as it is simpler, but I would like to keep similar structure across problems.

The way we thought of it was to have an example where users can define their own cost function. Note that some cost functions introduce new variables (for MLE it is the standard deviations of each data series) which means that the example above would need to take those separately as they are not part of the solution directly.

**Comment by muhammedsogut** [comment]
I think it might create problems using same cost function class for various `OptimisationProblem` we have. Because `OCPBalance` does not take `variables_optimise` as input. Some variables are problem specific and it may make more sense to request them when the problem is defined.

**Comment by tinosulzer** [comment]
Yeah, what I'm referring to would be specific to the `DataFit` class. I had not thought of the MLE case though. Maybe the existing API is more appropriate
How is the sd calculated?

**Comment by brosaplanella** [comment]
That's not implemented yet but the idea is that the SD of each series is an additional parameter to be fit. The idea will be that, when assembling the optimisation problem, the code puts together the variables to optimise arising from the problem and those arising from the cost function.
